File: SCRIPTS/data/websocket_client.js
// ============================================================
// websocket_client.js - Auto-Reconnect WebSocket (K130)
// Exponential backoff + heartbeat + reconnect limit.
// SignalR veya native WS destekler.
// ============================================================

const WebSocket = require('ws');
const logger = require('../core/logger');

class ReconnectingWebSocket {
    constructor(url, options = {}) {
        this.url = url;
        this.protocols = options.protocols || [];
        this.maxAttempts = options.maxAttempts || 10;
        this.baseDelayMs = options.baseDelayMs || 1000;
        this.maxDelayMs = options.maxDelayMs || 30000;
        this.heartbeatIntervalMs = options.heartbeatIntervalMs || 30000;
        this.heartbeatMsg = options.heartbeatMsg || JSON.stringify({ type: 'ping' });
        this.reconnectDecay = options.reconnectDecay || 2;

        this.ws = null;
        this.attempts = 0;
        this.reconnectTimer = null;
        this.heartbeatTimer = null;
        this.ready = false;
        this._listeners = { open: [], message: [], close: [], error: [] };
        this._queue = [];

        this._connect();
    }

    _connect() {
        try {
            logger.info(`[WS] Baglaniliyor: ${this.url}`);
            this.ws = new WebSocket(this.url, this.protocols);

            this.ws.on('open', () => {
                this.attempts = 0;
                this.ready = true;
                logger.info(`[WS] Baglandi: ${this.url}`);
                this._startHeartbeat();
                this._flushQueue();
                this._emit('open');
            });

            this.ws.on('message', (data) => {
                this._emit('message', data);
            });

            this.ws.on('close', (code, reason) => {
                this.ready = false;
                this._stopHeartbeat();
                logger.warn(`[WS] Kapandi: ${code} ${reason}`);
                this._emit('close', { code, reason });
                this._scheduleReconnect();
            });

            this.ws.on('error', (err) => {
                logger.error(`[WS] Hata: ${err.message}`);
                this._emit('error', err);
            });
        } catch (err) {
            logger.error(`[WS] Baglanti hatasi: ${err.message}`);
            this._scheduleReconnect();
        }
    }

    _scheduleReconnect() {
        if (this.attempts >= this.maxAttempts) {
            logger.error(`[WS] Max deneme asildi (${this.maxAttempts}). Baglanti birakildi.`);
            return;
        }
        const delay = Math.min(this.baseDelayMs * Math.pow(this.reconnectDecay, this.attempts), this.maxDelayMs);
        this.attempts++;
        logger.info(`[WS] ${delay}ms sonra yeniden baglanilacak (deneme ${this.attempts}/${this.maxAttempts})`);
        this.reconnectTimer = setTimeout(() => this._connect(), delay);
    }

    _startHeartbeat() {
        this.heartbeatTimer = setInterval(() => {
            if (this.ready) this.send(this.heartbeatMsg);
        }, this.heartbeatIntervalMs);
    }

    _stopHeartbeat() {
        if (this.heartbeatTimer) { clearInterval(this.heartbeatTimer); this.heartbeatTimer = null; }
    }

    _flushQueue() {
        while (this._queue.length > 0 && this.ready) {
            const msg = this._queue.shift();
            this.ws.send(msg);
        }
    }

    send(data) {
        const msg = typeof data === 'string' ? data : JSON.stringify(data);
        if (this.ready) {
            this.ws.send(msg);
        } else {
            this._queue.push(msg);
        }
    }

    on(event, handler) {
        if (!this._listeners[event]) this._listeners[event] = [];
        this._listeners[event].push(handler);
    }

    _emit(event, data) {
        for (const h of this._listeners[event] || []) {
            try { h(data); } catch (e) { logger.error(`[WS] Listener hatasi: ${e.message}`); }
        }
    }

    close() {
        this._stopHeartbeat();
        if (this.reconnectTimer) { clearTimeout(this.reconnectTimer); this.reconnectTimer = null; }
        if (this.ws) { this.ws.close(); this.ws = null; }
        this.attempts = this.maxAttempts + 1; // Prevent reconnect
    }

    getStatus() {
        return {
            url: this.url,
            ready: this.ready,
            attempts: this.attempts,
            queueSize: this._queue.length,
        };
    }
}

module.exports = ReconnectingWebSocket;

File: PYTHON/data/websocket_client.py
"""
websocket_client.py — Auto-Reconnect WebSocket Client (K130)
Exponential backoff + heartbeat + reconnect limit.
"""
import asyncio
import random
import time
from typing import Callable, Optional


class ReconnectingWebSocket:
    """
    WebSocket baglanti yonetimi: disconnect oldugunda otomatik reconnect.
    """

    def __init__(
        self,
        url: str,
        on_message: Callable,
        on_open: Optional[Callable] = None,
        on_close: Optional[Callable] = None,
        max_attempts: int = 10,
        base_delay_ms: float = 1000,
        max_delay_ms: float = 30000,
        heartbeat_interval_sec: float = 30,
        reconnect_decay: float = 2,
    ):
        self.url = url
        self.on_message = on_message
        self.on_open = on_open
        self.on_close = on_close
        self.max_attempts = max_attempts
        self.base_delay_ms = base_delay_ms
        self.max_delay_ms = max_delay_ms
        self.heartbeat_interval_sec = heartbeat_interval_sec
        self.reconnect_decay = reconnect_decay
        self.attempts = 0
        self.ready = False
        self._ws = None
        self._reconnect_task = None
        self._heartbeat_task = None

    async def connect(self):
        """Baglantiyi baslat."""
        await self._connect()

    async def _connect(self):
        import websockets
        try:
            self._ws = await websockets.connect(self.url)
            self.attempts = 0
            self.ready = True
            if self.on_open:
                self.on_open()
            self._start_heartbeat()
            await self._listen()
        except Exception:
            self.ready = False
            if self.on_close:
                self.on_close()
            await self._schedule_reconnect()

    async def _listen(self):
        import websockets
        try:
            async for message in self._ws:
                self.on_message(message)
        except websockets.exceptions.ConnectionClosed:
            self.ready = False
            if self.on_close:
                self.on_close()
            await self._schedule_reconnect()

    async def _schedule_reconnect(self):
        if self.attempts >= self.max_attempts:
            return
        delay = min(self.base_delay_ms * (self.reconnect_decay ** self.attempts), self.max_delay_ms)
        self.attempts += 1
        await asyncio.sleep(delay / 1000)
        await self._connect()

    def _start_heartbeat(self):
        async def heartbeat():
            while self.ready and self._ws:
                try:
                    await self._ws.send('{"type":"ping"}')
                except Exception:
                    break
                await asyncio.sleep(self.heartbeat_interval_sec)
        asyncio.create_task(heartbeat())

    async def send(self, payload: str):
        if self.ready and self._ws:
            await self._ws.send(payload)

    async def close(self):
        self.ready = False
        if self._ws:
            await self._ws.close()
